crate_universe/tests/renderer.rs

```diff
@@ -0,0 +1,334 @@
+}
+
+#[test]
+fn single_http_archive() {
+    let renderer = Renderer::new(
+        RenderConfig {
+            repo_rule_name: String::from("rule_prefix"),
+            repository_template: String::from(
+                "https://crates.io/api/v1/crates/{crate}/{version}/download",
+            ),
+            rules_rust_workspace_name: String::from("rules_rust"),
+        },
+        String::from("598"),
+        vec![lazy_static_crate_context(false)],
+        Dependencies {
+            normal: BTreeMap::new(),
+            build: BTreeMap::new(),
+            dev: BTreeMap::new(),
+        },
+        BTreeMap::new(),
+    );
+
+    let mut output = Vec::new();
+
+    renderer
+        .render_workspaces(&mut output)
+        .expect("Error rendering");
+
+    let output = String::from_utf8(output).expect("Non-UTF8 output");
+
+    // TODO: Don't unconditionally load new_git_repository and http_archive
+    let expected_repository_rule = format!(
+        r###"# rules_rust crate_universe file format 1
+# config hash 598
+
+load("@bazel_tools//tools/build_defs/repo:git.bzl", "new_git_repository")
+
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+
+def pinned_rust_install():
+    http_archive(
+        name = "rule_prefix__lazy_static__1_4_0",
+        build_file_content = """{}""",
+        sha256 = "e2abad23fbc42b3700f2f279844dc832adb2b2eb069b2df918f455c4e18cc646",
+        strip_prefix = "lazy_static-1.4.0",
+        type = "tar.gz",
+        url = "https://crates.io/api/v1/crates/lazy_static/1.4.0/download",
+    )
+"###,
+        LAZY_STATIC_BUILD_FILE_CONTENT
+    );
+
+    assert_eq!(output, expected_repository_rule);
+}
+
+#[test]
+fn crate_helper_function_mapping() {
+    let crate_context = lazy_static_crate_context(false);
+
+    let normal_deps = btreemap! {
+        crate_context.pkg_name.clone() => crate_context.pkg_version.clone(),
+    };
+
+    let renderer = Renderer::new(
+        RenderConfig {
+            repo_rule_name: String::from("rule_prefix"),
+            repository_template: String::from(
+                "https://crates.io/api/v1/crates/{crate}/{version}/download",
+            ),
+            rules_rust_workspace_name: String::from("rules_rust"),
+        },
+        String::from("598"),
+        vec![crate_context],
+        Dependencies {
+            normal: normal_deps,
+            build: BTreeMap::new(),
+            dev: BTreeMap::new(),
+        },
+        btreemap! {
+            String::from("//some:Cargo.toml") => btreeset!{ String::from("lazy_static") },
+        },
+    );
+
+    let mut output = Vec::new();
+
+    renderer
+        .render_helper_functions(&mut output)
+        .expect("Error rendering");
+
+    let output = String::from_utf8(output).expect("Non-UTF8 output");
+
+    // TODO: Have a single function with kwargs to enable each kind of dep, rather than multiple functions.
+    let expected_repository_rule = r###"CRATE_TARGET_NAMES = {
```

Comment:
Feeling nervous about the volume of dumped text in the tests right now. I don't think this is a blocker for this PR, but I think it'd be better to implement https://github.com/bazelbuild/rules_rust/issues/653 and have more asserts around data instead of massive string comparison.

Reply:
Oh yeah, I think that everything _except_ the renderer should be tested using much more data-y ways, but I think the renderer fundamentally kind of _has_ to be checking raw strings against each other.

For the renderer tests themselves, making the things being tested be much more targeted (e.g. single crates), and hermetic (hard-coded inputs), and potentially using better assertions libraries, helps here, but at the end of the day these tests test what we're really outputting the right thing. I also find it pretty useful to be able to look at the diffs in PRs to see what "really" changed. I'm hoping that these small number of better tests will be sufficiently maintainable, and that we'll delete the huge unhermetic ones when we have suitable targeted replacements for each pipeline stage :)

Reply:
Ah, yeah, sorry, I didn't realize I'd put that comment in the renderer code. I agree the renderer should be the place to do these kinds of tests. This is my first time really looking at the tests so just wanted to reiterate that issue.